server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.search.TopDocs;
+import org.elasticsearch.action.search.SearchPhaseController.TopDocsStats;
+import org.elasticsearch.common.io.stream.DelayableWriteable;
+import org.elasticsearch.common.io.stream.DelayableWriteable.Serialized;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.lucene.search.TopDocsAndMaxScore;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.search.aggregations.InternalAggregation.ReduceContextBuilder;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.query.QuerySearchResult;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.action.search.SearchPhaseController.mergeTopDocs;
+import static org.elasticsearch.action.search.SearchPhaseController.setShardIndex;
+
+/**
+ * A {@link ArraySearchPhaseResults} implementation that incrementally reduces aggregation results
+ * as shard results are consumed.
+ * This implementation can be configured to batch up a certain amount of results and reduce
+ * them asynchronously in the provided {@link Executor} iff the buffer is exhausted.
+ */
+class QueryPhaseResultConsumer extends ArraySearchPhaseResults<SearchPhaseResult> {
```

Comment:
Was this class only extracted from SearchPhaseController or what changes have been made to it? Maybe for easier reviewing it could have stayed where it was, at least initially? Would that be possible?

Reply:
We could but the change is substantial so I don't think it would help. This PR is a complete rewrite of the `QueryPhaseResultConsumer`.

Reply:
Yeah. This looks so much bigger than before and I don't think it would have made review a ton easier to keep it where it was.